Sources/NIOTestUtils/EventCounterHandler.swift

```diff
@@ -0,0 +1,363 @@
+/// `EventCounterHandler` is mostly useful in unit tests to validate other `ChannelHandler`'s behaviour.
+///
+/// - note: Contrary to most `ChannelHandler`s, all of `EventCounterHandler`'s API is thread-safe meaning that you can
+///         query the events received from any thread.
+public class EventCounterHandler: ChannelDuplexHandler {
+    public typealias InboundIn = Any
+    public typealias InboundOut = Any
+    public typealias OutboundIn = Any
+    public typealias OutboundOut = Any
+
+    public struct ValidityError: Error {
+        public var reason: String
+
+        init(_ reason: String) {
+            self.reason = reason
+        }
+    }
+
+    private let _channelRegisteredCalls = Atomic<Int>(value: 0)
+    private let _channelUnregisteredCalls = Atomic<Int>(value: 0)
+    private let _channelActiveCalls = Atomic<Int>(value: 0)
+    private let _channelInactiveCalls = Atomic<Int>(value: 0)
+    private let _channelReadCalls = Atomic<Int>(value: 0)
+    private let _channelReadCompleteCalls = Atomic<Int>(value: 0)
+    private let _channelWritabilityChangedCalls = Atomic<Int>(value: 0)
+    private let _userInboundEventTriggeredCalls = Atomic<Int>(value: 0)
+    private let _errorCaughtCalls = Atomic<Int>(value: 0)
+    private let _registerCalls = Atomic<Int>(value: 0)
+    private let _bindCalls = Atomic<Int>(value: 0)
+    private let _connectCalls = Atomic<Int>(value: 0)
+    private let _writeCalls = Atomic<Int>(value: 0)
+    private let _flushCalls = Atomic<Int>(value: 0)
+    private let _readCalls = Atomic<Int>(value: 0)
+    private let _closeCalls = Atomic<Int>(value: 0)
+    private let _triggerUserOutboundEventCalls = Atomic<Int>(value: 0)
+
+    public init() {}
+
+    /// Returns the number of `channelRegistered` events seen so far in the `ChannelPipeline`.
+    public var channelRegisteredCalls: Int {
+        return self._channelRegisteredCalls.load()
+    }
+
+    /// Returns the number of `channelUnregistered` events seen so far in the `ChannelPipeline`.
+    public var channelUnregisteredCalls: Int {
+        return self._channelUnregisteredCalls.load()
+    }
+
+    /// Returns the number of `channelActive` events seen so far in the `ChannelPipeline`.
+    public var channelActiveCalls: Int {
+        return self._channelActiveCalls.load()
+    }
+
+    /// Returns the number of `channelInactive` events seen so far in the `ChannelPipeline`.
+    public var channelInactiveCalls: Int {
+        return self._channelInactiveCalls.load()
+    }
+
+    /// Returns the number of `channelRead` events seen so far in the `ChannelPipeline`.
+    public var channelReadCalls: Int {
+        return self._channelReadCalls.load()
+    }
+
+    /// Returns the number of `channelReadComplete` events seen so far in the `ChannelPipeline`.
+    public var channelReadCompleteCalls: Int {
+        return self._channelReadCalls.load()
```

Comment:
This is returning the wrong atomic value. Looks like you may need a better unit test for this one.

Reply:
thank you! I did now check all individual events.